### PR TITLE
Omit data Key in Error Responses When No Data is Present

### PR DESCRIPTION
**Pull request by surajit-zs**
This PR updates the response format for error cases, ensuring that the data key is omitted from the response body when there is no data to return. Previously, responses included a data key with a null value, which added unnecessary clutter to the error messages. The changes aim to produce cleaner and more concise error responses.

Added an `isNil` utility function to determine if an interface contains a nullable type with a nil value. This helps in accurately deciding whether to include the data key.

- Closes: #912

### CODE REVIEW

**Review by aryanmehrotra — DISMISSED** [review]

File: pkg/gofr/http/responder.go [hunk 1]
```diff
@@ -35,9 +36,11 @@ func (r Responder) Respond(data interface{}, err error) {
...
-		resp = response{
-			Data:  v,
-			Error: errorObj,
+		resp = response{Data: data, Error: errorObj}
+
+		// handling where an interface contains a nullable type with a nil value.
+		if isNil(data) {
+			resp = response{Error: errorObj}
 		}
 	}
```
Comment (ccoVeille) [inline pkg/gofr/http/responder.go:45]: I'm unsure to understand why the omitempty is not enough?

File: pkg/gofr/http/responder.go [hunk 2]
```diff
@@ -35,9 +36,11 @@ func (r Responder) Respond(data interface{}, err error) {
...
 	default:
-		resp = response{
-			Data:  v,
-			Error: errorObj,
+		resp = response{Data: data, Error: errorObj}
+
+		// handling where an interface contains a nullable type with a nil value.
+		if isNil(data) {
+			resp = response{Error: errorObj}
 		}
```
Comment (ccoVeille) [inline pkg/gofr/http/responder.go:44]: The code is a bit strange to read

I would expect something like


```suggestion
		resp = response{Data: data, Error: errorObj}

		// handling where an interface contains a nullable type with a nil value.
		if isNil(data) {
			resp.Data = nil
 		}
```

or


```suggestion
		resp = response{Error: errorObj}

		if !isNil(data) {
			resp.Data = data
 		}
```

Comment (surajit-zs) [reply pkg/gofr/http/responder.go:44]: @ccoVeille  sure we can go with 			
```go
resp.Data = nil
```

File: pkg/gofr/http/responder.go (hunk 1 above)
Comment (surajit-zs) [reply pkg/gofr/http/responder.go:45]: image

In this case, it is not working. 

I'm not entirely sure why this is happening and might need to read more about it.

File: pkg/gofr/http/responder.go (hunk 2 above)
Comment (ccoVeille) [reply pkg/gofr/http/responder.go:44]: Maybe I wasn't able to explain correctly.

I'll suggest something else

Comment (ccoVeille) [inline pkg/gofr/http/responder.go:44]: This code would be simpler I think
	
```suggestion
		// handling where an interface contains a nullable type with a nil value.
		if isNil(data) {
			data = nil
 		}
		resp = response{Data: data, Error: errorObj}
```

**Review by ccoVeille — APPROVED** [review]

**Comment by srijan-27** [comment]
This PR handles case where the response is `nil` (which is working fine). But what about these cases:
- If the response is empty and error is also present, then the body contains both error and data (empty data)
- If the response is valid struct with error, then the body contains both error and data. In this case I believe the statusCode should be `206`, but its `500`.

FYI: @surajit-zs, @vipul-rawat, @Umang01-hash, @aryanmehrotra

**Comment by vipul-rawat** [comment]
> This PR handles cases where the response is `nil` (which is working fine). But what about these cases:
> 
> * If the response is empty and error is also present, then the body contains both error and data (empty data)
> * If the response is valid struct with error, then the body includes both error and data. In this case I believe the statusCode should be `206`, but its `500`.
> 
> FYI: @surajit-zs, @vipul-rawat, @Umang01-hash, @aryanmehrotra

I think the data field should be present for empty responses (zero values) and we should not modify the user-sent response. For `206` we can create another issue to solve this status code problem.

**Comment by srijan-27** [comment]
created a ticket: #938

**Review by srijan-27 — APPROVED** [review]

**Review by aryanmehrotra — APPROVED** [review]

**Pull request merged**